packages/ab-testing/src/index.ts

```diff
@@ -1,9 +1,91 @@
-// add another package with `yo lerna-typescript:package @myscope/my-new-package` instead of `lerna create`
+import 'ts-polyfill/lib/es2019-object';
+import createHash from 'create-hash';
+import crc32 from 'fast-crc32c';
 
-export function greet(name: string): string {
-    return `ab-testing says: hello to ${name}`;
+interface ForceInclude {
+    [hashedKey: string]: string[];
 }
 
-export function add(a: number, b: number): number {
-    return a + b;
+interface Cohort {
+    name: string;
+    allocation?: [number, number][];
+    force_include?: ForceInclude;
 }
+
+interface Experiment {
+    name: string;
+    cohorts: Cohort[];
+}
+
+export interface ABTestingConfig {
+    salt: string;
+    version: '1.0';
+    experiments: Experiment[];
+}
+
+export interface UserProfile {
+    user_id: number;
+    user_type?: string;
+    email?: string;
+    email_domain?: string;
+}
+
+function getHashedValue(salt: string, sourceValue: unknown): string {
+    const hash = createHash('sha256');
+    hash.update(salt);
+    hash.update(String(sourceValue));
+    return hash.digest('hex');
+}
+
+function hashObject(obj: object, salt: string): { [s: string]: string } {
+    return Object.fromEntries(
+        Object.entries(obj).map(([key, value]) => [
+            getHashedValue(salt, key),
+            getHashedValue(salt, value),
+        ])
+    );
+}
+
+function getModuloValue(experiment: string, userId: number): number {
+    return crc32.calculate(String(userId), crc32.calculate(experiment)) % 100;
+}
+
+export class Experiments {
+    config: ABTestingConfig;
+    profile: UserProfile;
+    hashedProfile: { [s: string]: string };
+    constructor(
+        config: ABTestingConfig,
+        profile: UserProfile,
+        hashProfile: typeof hashObject = hashObject
+    ) {
+        this.config = config;
+        this.profile = profile;
+        this.hashedProfile = hashProfile(profile as object, config.salt);
+    }
+
+    getCohort = (experimentName: string): string => {
+        const experimentConfig = this.config.experiments.find(e => e.name === experimentName);
+        for (const cohort of experimentConfig?.cohorts || []) {
+            for (const [hashedProfileKey, hashedProfileVal] of Object.entries(this.hashedProfile)) {

```

Comment:
We should loop `force_include` instead of the hashed profile. This will be faster as we'll only loop keys we know are relevant.